### PR TITLE
Add support for CRAM, check contigs name

### PR DESCRIPTION
**Pull request by LouisLeNezet**
Here we add support for CRAM file as input.
The contigs names are now check at workflow level for all input (bam and vcf).
Fix glimpse2 file naming.
Remove samples in one step with removing multiallelics records.
Publish only once the reference panel.
Moove samtools merge to BAM_REGION sbwf and fix renaming.

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/phaseimpute/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/phaseimpute _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: conf/steps/panel_prep.config [hunk 1]
```diff
@@ -28,7 +28,7 @@ process {
     }
 
     withName: 'NFCORE_PHASEIMPUTE:PHASEIMPUTE:VCF_CHR_CHECK:VCF_CHR_RENAME:GAWK' {
-        ext.args   = 'BEGIN {FS="\\t"} NR==1 { if ($1 ~ /^chr/) { col1=""; col2="chr" } else { col1="chr"; col2="" } } { sub(/^chr/, "", $1); if ($1 ~ /^[0-9]+|[XYMT]$/) print col1$1, col2$1; else print $1, $1 }'
+        ext.args2  = 'BEGIN {FS="\\t"} NR==1 { if ($1 ~ /^chr/) { col1=""; col2="chr" } else { col1="chr"; col2="" } } { sub(/^chr/, "", $1); if ($1 ~ /^[0-9]+|[XYMT]$/) print col1$1, col2$1; else print $1, $1 }'
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply conf/steps/panel_prep.config:31]: Because this is the dedicated argument in GAWK for the script to go.
ext.args is for other line command argument.

File: conf/test_panelprep_fullchr.config [hunk 2]
```diff
@@ -20,10 +20,14 @@ params {
     max_time   = '4.h'
 
     // Genome references
-    fasta  = "https://raw.githubusercontent.com/nf-core/test-datasets/phaseimpute/data/reference_genome/21_22/hs38DH.chr21_22.fa"
+    fasta = "https://raw.githubusercontent.com/nf-core/test-datasets/phaseimpute/data/reference_genome/21_22/hs38DH.chr21_22.fa"
     panel  = "${projectDir}/tests/csv/panel_fullchr.csv"
     phased = false
-
+    input        = "${projectDir}/tests/csv/sample_sim_full.csv"
+    input_region = "${projectDir}/tests/csv/region_full.csv"
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply conf/test_panelprep_fullchr.config:27]: True, it should normally extract the full regions directly from the fai file.
I've just tried and this seems to work.

File: modules/nf-core/glimpse2/phase/main.nf [hunk 3]
```diff
@@ -18,7 +18,7 @@ process GLIMPSE2_PHASE {
         'biocontainers/glimpse-bio:2.0.0--hf340a29_0' }"
 
     input:
-        tuple val(meta) , path(input), path(input_index), path(samples_file), val(input_region), val(output_region), path(reference), path(reference_index), path(map)
+        tuple val(meta) , path(input), path(input_index), path(samples_file), val(input_region), val(output_region), path(reference), path(reference_index), path(map), path(bamlist)
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply modules/nf-core/glimpse2/phase/main.nf:21]: This was to correctly set up the glimpse2_phase process in the case of multiple input as an array.
We could also perform a test on the input parameter to see if it's an array and in this case test that all of them a ".bam" or ".cram"

File: modules/nf-core/samtools/reheader/main.nf [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    tuple val(meta), path("*.bam"), emit: bam
+    path "versions.yml"           , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def cmd_reheader = cmd ? "-c '$cmd'" : ""
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply modules/nf-core/samtools/reheader/main.nf:23]: No, as it needs to be changed depending on the presence / absence of the chr prefix in the file.

File: subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf [hunk 5]
```diff
@@ -218,8 +219,8 @@ workflow PIPELINE_INITIALISATION {
         ch_hap_legend = Channel.fromSamplesheet("posfile")
                 .map { meta, vcf, index, txt, hap, legend -> [meta, hap, legend] }
     } else {
-        ch_posfile = [[],[]]
-        ch_hap_legend   = Channel.empty()
+        ch_posfile = Channel.of([[chr:null], [], [], []])
+        ch_hap_legend   = Channel.of([[chr:null],[],[]])
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf:223]: Setting all default output to channels ensure that the following operations will always be done on channels.
The chr key in the meta map is just to allow an easier debugging when a join doesn't hapen as expected.

File: nf-test.config [hunk 6]
```diff
@@ -9,6 +12,10 @@ config {
     configFile "./nextflow.config"
     configFile "tests/config/nf-test.config"
 
+    stage {
+        copy "./nextflow_schema.json"
+    }
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply nf-test.config:17]: This stage allow to add the `nextflow_shema.json` to the asset during an nf-test. It allows to test the `utils_nfcore_phaseimpute_pipeline/main.nf`

File: nf-test.config [hunk 7]
```diff
@@ -1,6 +1,9 @@
 config {
+    plugins {
+        load "nft-bam@0.2.0"
+    }
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply nf-test.config:4]: Yep I also aim in the future to use the `nft_vcf` for variant calling format files.

File: main.nf [hunk 8]
```diff
@@ -55,6 +58,24 @@ workflow NFCORE_PHASEIMPUTE {
...
 
+    //  Check input files for contigs names consistency
+    lst_chr = ch_regions.map { it[0].chr }
+        .unique()
+        .collect()
+        .toList()
+
+    CHRCHECK_INPUT(ch_input.combine(lst_chr))
+    ch_input = CHRCHECK_INPUT.out.output
+    ch_versions = ch_versions.mix(CHRCHECK_INPUT.out.versions)
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply main.nf:69]: Versions are exported here, as the `ch_versions` is then passed to the phaseimpute workflow.

File: main.nf [hunk 9]
```diff
@@ -55,6 +58,24 @@ workflow NFCORE_PHASEIMPUTE {
     ch_input_simulate       = Channel.empty()
     ch_input_validate       = Channel.empty()
 
+    //  Check input files for contigs names consistency
+    lst_chr = ch_regions.map { it[0].chr }
+        .unique()
+        .collect()
+        .toList()
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply main.nf:65]: Done.

File: main.nf [hunk 10]
```diff
@@ -55,6 +58,24 @@ workflow NFCORE_PHASEIMPUTE {
     ch_input_simulate       = Channel.empty()
     ch_input_validate       = Channel.empty()
 
+    //  Check input files for contigs names consistency
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply main.nf:61]: We need to discuss it a bit if it's a subworkflow or a workflow.
I will get in touch with you by slack on that.

File: workflows/phaseimpute/main.nf [hunk 11]
```diff
@@ -103,22 +102,23 @@ workflow PHASEIMPUTE {
...
         if (params.input_region) {
             // Split the bam into the regions specified
             BAM_REGION(ch_input_sim, ch_region, ch_fasta)
             ch_versions  = ch_versions.mix(BAM_REGION.out.versions)
             ch_input_sim = BAM_REGION.out.bam_region
+            ch_input_validate_truth = BAM_REGION.out.bam_region
         }
 
+        // Compute coverage of input files
+        SAMTOOLS_COVERAGE_TRT(ch_input_validate_truth, ch_fasta)
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply workflows/phaseimpute/main.nf:118]: You right it is now renamed `INP` (input) and `DWN` (downsampled)

File: workflows/chrcheck/main.nf [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        ch_output = chr_bam_disjoint.no_rename
+            .mix(chr_vcf_disjoint.no_rename)
+            .mix(ch_bam_renamed)
+            .mix(ch_vcf_renamed)
+    emit:
+        output   = ch_output             // [ [id], file, index ]
+        versions = ch_versions           // channel: [ versions.yml ]
+}
+
+def checkChr(ch_chr, ch_input){
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply workflows/chrcheck/main.nf:71]: I do not know.
I personnaly find it convenient to have them in the same script for debugging.

File: workflows/chrcheck/main.nf [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+workflow CHRCHECK {
+    take:
+        ch_input // [[id], file, index, [chr]]
+
+    main:
+        ch_versions = Channel.empty()
+        // Split the input between VCF and BAM files
+        ch_input = ch_input.branch{
+            bam: it[1] =~ 'bam|cram|sam'
+            vcf: it[1] =~ 'vcf|bcf'
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply workflows/chrcheck/main.nf:23]: Another more solution would be to test if the given channel is empty or not and process it in this case.
What do you think ?

File: subworkflows/local/vcf_chr_rename_bcftools/main.nf [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    ch_rename_file = ch_vcf
+        .collectFile{ meta, vcf, index, diff, prefix ->
+            chr = ""
+            if (prefix == "chr") {
+                for (i in diff) {
+                    chr += "${i} chr${i}\n"
+                }
+            } else if (prefix == "nochr") {
+                for (i in diff) {
+                    chr += "${i} ${i.replace('chr', '')}\n"
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply subworkflows/local/vcf_chr_rename_bcftools/main.nf:23]: This is effectively prettier and more Nextflow style !

File: main.nf [hunk 15]
```diff
@@ -55,6 +58,22 @@ workflow NFCORE_PHASEIMPUTE {
...
+    lst_chr = ch_regions.map { it[0].chr }
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply main.nf:62]: I agree, I will add a check in the initialisation phase.

[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply main.nf:62]: When the user use a fasta without chr and specify chr in the regions samplesheet this will produce an error.

File: subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf [hunk 16]
```diff
@@ -229,9 +230,42 @@ workflow PIPELINE_INITIALISATION {
...
-        ch_chunks = [[],[]]
+        ch_chunks = Channel.of([[chr:null],[]])
     }
 
+    //
+    // Check contigs name in different meta map
+    //
+    // Collect all chromosomes names in all different inputs
+    chr_ref = ch_ref_gen.map { meta, fasta, fai -> [fai.readLines()*.split('\t').collect{it[0]}] }
+    chr_regions = ch_regions.map { meta, region -> [meta.chr] }
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf:241]: I agree. Is it better ?

File: subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf [hunk 17]
```diff
@@ -229,9 +230,42 @@ workflow PIPELINE_INITIALISATION {
...
+        .collect()
+        .toList()
+    chr_hap_legend = ch_hap_legend.map { meta, hap, legend -> [meta.chr] }
+        .collect()
+        .toList()
+    chr_posfile = ch_posfile.map { meta, sites, index, posfile -> [meta.chr] }
+        .collect()
+        .toList()
+
+    // Check that the chromosomes names that will be used are all present in different inputs
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply subworkflows/local/utils_nfcore_phaseimpute_pipeline/main.nf:260]: I don't think so.
The aim in the end is to only process the chr_regions. If other contigs are present in the input files they won't be used.

File: README.md [hunk 18]
```diff
@@ -40,19 +40,20 @@ First, prepare a samplesheet with your input data that looks as follows:
 `samplesheet.csv`:
 
 ```csv
-sample,bam,bai
+sample,file,index
 1_BAM_1X,/path/to/.bam,/path/to/.bai
+1_CRAM_1X,/path/to/.cram,/path/to/.crai
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply README.md:45]: For the moment all of them need to be the same.
This would need a bit of extra work if we want both to be supported.

File: CHANGELOG.md [hunk 19]
```diff
@@ -35,6 +35,7 @@ Initial release of nf-core/phaseimpute, created with the [nf-core](https://nf-co
 - [#86](https://github.com/nf-core/phaseimpute/pull/86) - Move `bcftools_convert` to `vcf_sites_extract_bcftools`.
 - [#88](https://github.com/nf-core/phaseimpute/pull/88) - Improve multiqc report with more information.
 - [#91](https://github.com/nf-core/phaseimpute/pull/91) - Update metro map with all steps and remove deprecated ones.
+- [#93](https://github.com/nf-core/phaseimpute/pull/93) - Add support for CRAM file. Check contigs name at workflow level for BAM and VCF. Samples remove with multiallelics records. Samtools merge in BAM_REGION sbwf. Fix glimpse2_phase output file names. Fix fai combination to fasta.
```
Comment (maxulysse) [inline CHANGELOG.md:38]:
```suggestion
- [#93](https://github.com/nf-core/phaseimpute/pull/93) - Add support for CRAM file.
- [#93](https://github.com/nf-core/phaseimpute/pull/93) - Check contigs name at workflow level for BAM and VCF.
- [#93](https://github.com/nf-core/phaseimpute/pull/93) - Samples remove with multiallelics records.
- [#93](https://github.com/nf-core/phaseimpute/pull/93) - Samtools merge in BAM_REGION sbwf.
- [#93](https://github.com/nf-core/phaseimpute/pull/93) - Fix glimpse2_phase output file names.
- [#93](https://github.com/nf-core/phaseimpute/pull/93) - Fix fai combination to fasta.
```

That sounds like 5 different fixes in the 1 PR

File: assets/schema_input.json [hunk 20]
```diff
@@ -15,13 +15,13 @@
...
             "file": {
                 "type": "string",
-                "pattern": "^\\S+\\.(bam)|((vcf|bcf)(\\.gz))?$",
-                "errorMessage": "BAM, VCF or BCF file must be provided, cannot contain spaces and must have extension '.bam' or '.vcf', '.bcf' with optional '.gz' extension"
+                "pattern": "^\\S+\\.(bam|cram)|((vcf|bcf)(\\.gz))?$",
+                "errorMessage": "BAM, CRAM, VCF or BCF file must be provided, cannot contain spaces and must have extension '.bam', '.cram' or '.vcf', '.bcf' with optional '.gz' extension"
             },
             "index": {
-                "errorMessage": "Input file index must be provided, cannot contain spaces and must have extension '.bai', '.tbi' or '.csi'",
+                "errorMessage": "Input file index must be provided, cannot contain spaces and must have extension '.bai', 'crai', '.tbi' or '.csi'",
```
Comment (maxulysse) [inline assets/schema_input.json:22]: Oooh,
so index could be bam/cram index or vcf index

File: conf/steps/validation.config [hunk 21]
```diff
@@ -66,11 +66,10 @@ process {
             mode: params.publish_dir_mode,
             saveAs: { filename -> filename.equals('versions.yml') ? null : filename }
         ]
-        tag = {"${meta.id} ${meta.panel}"}
+        tag = {"${meta.id} ${meta.panel} ${meta.tools}"}
```
Comment (maxulysse) [inline conf/steps/validation.config:69]: <3

File: conf/test_all_fullchr.config [hunk 22]
```diff
@@ -23,7 +23,10 @@ params {
     fasta  = "https://raw.githubusercontent.com/nf-core/test-datasets/phaseimpute/data/reference_genome/21_22/hs38DH.chr21_22.fa"
     panel  = "${projectDir}/tests/csv/panel_fullchr.csv"
     phased = false
-
+    input        = "${projectDir}/tests/csv/sample_sim_full.csv"
```
Comment (maxulysse) [inline conf/test_all_fullchr.config:26]:
```suggestion
    input  = "${projectDir}/tests/csv/sample_sim_full.csv"
```

File: docs/images/metro/.$MetroMap.xml.bkp
Comment (maxulysse) [inline docs/images/metro/.$MetroMap.xml.bkp:1]: is this file useful?

File: modules/local/bam_chr_extract/environment.yml [hunk 23]
```diff
@@ -0,0 +1,7 @@
+name: bam_chr_extract
+channels:
+  - conda-forge
+  - bioconda
+  - defaults
+dependencies:
+  - bioconda::samtools=1.20
```
Comment (maxulysse) [inline modules/local/bam_chr_extract/environment.yml:7]: I usually pin htslib as well

File: subworkflows/local/vcf_normalize_bcftools/main.nf [hunk 24]
```diff
@@ -2,9 +2,7 @@ include { BCFTOOLS_NORM                         } from '../../../modules/nf-core
 include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_1    } from '../../../modules/nf-core/bcftools/index'
 include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_2    } from '../../../modules/nf-core/bcftools/index'
 include { BCFTOOLS_INDEX as BCFTOOLS_INDEX_3    } from '../../../modules/nf-core/bcftools/index'
```
Comment (maxulysse) [inline subworkflows/local/vcf_normalize_bcftools/main.nf:4]: not super fond of the _1, _2 and _3 names for the modules.
Any better ideas?

File: workflows/phaseimpute/main.nf [hunk 25]
```diff
@@ -103,32 +102,35 @@ workflow PHASEIMPUTE {
     if (params.steps.split(',').contains("simulate") || params.steps.split(',').contains("all")) {
         // Test if the input are all bam files
         getAllFilesExtension(ch_input_sim)
-            .map{ if (it != "bam") {
-                error "All input files must be in BAM format to perform simulation"
+            .map{ if (it != "bam" & it != "cram") {
+                error "All input files must be in the same format, either BAM or CRAM, to perform simulation"
```
Comment (maxulysse) [inline workflows/phaseimpute/main.nf:106]: Are you sure?
Did you try with a mixture of BAM/CRAM?

File: workflows/chrcheck/main.nf (hunk 12 above)
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply workflows/chrcheck/main.nf:80]: I think you're right as it is done like so in other pipeline.

File: assets/schema_input.json (hunk 20 above)
Comment (LouisLeNezet) [reply assets/schema_input.json:22]: Yeah should we test that the user correctly pass the expected index depending on the file ?

File: docs/images/metro/.$MetroMap.xml.bkp
Comment (LouisLeNezet) [reply docs/images/metro/.$MetroMap.xml.bkp:1]: Apparently no, I will delete it.

File: docs/usage.md [hunk 26]
```diff
@@ -377,13 +377,12 @@ The csv provided in `--posfile` must contain four columns [panel, chr, vcf, txt]
...
 - The second column [chr] is the chromosome corresponding to that file.
 - The third column [vcf], required in GLIMPSE1 imputation, is used for the computation of genotype likelihood in the preprocessing of glimpse1. In addition, this file is used in `--steps validate`.
-- The fourth column [txt] is a compressed tsv file containing the list of positions, unique to each chromosome.
 
 ```console
-panel,chr,vcf,txt
-1000G,chr1,posfile_chr1.vcf.gz,posfile_chr1.tsv.gz
-1000G,chr2,posfile_chr1.vcf.gz,posfile_chr2.tsv.gz
-1000G,chr3,posfile_chr1.vcf.gz,posfile_chr3.tsv.gz
+panel,chr,vcf
```
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply docs/usage.md:382]: It should but the whole documentation will change in the PR #96

File: subworkflows/local/vcf_normalize_bcftools/main.nf (hunk 24 above)
Comment (LouisLeNezet) [reply subworkflows/local/vcf_normalize_bcftools/main.nf:4]: This will be removed with the new version of all bcftools modules of nf-core with the native support of indexing in all modules.

File: workflows/phaseimpute/main.nf (hunk 25 above)
Comment (LouisLeNezet) [reply workflows/phaseimpute/main.nf:106]: Files need to be all bam or all cram for the moment.
Do we want to allow both at the same time ?

File: workflows/chrcheck/tests/function.nf.test
[the remark this answers is not in the document]
Comment (LouisLeNezet) [reply workflows/chrcheck/tests/function.nf.test:1]: This can be done but I just check directly in the function.nf.test the expected output.

File: modules/nf-core/glimpse2/phase/main.nf [hunk 27]
```diff
@@ -18,7 +18,7 @@ process GLIMPSE2_PHASE {
         'biocontainers/glimpse-bio:2.0.0--hf340a29_0' }"
 
     input:
-        tuple val(meta) , path(input), path(input_index), path(samples_file), val(input_region), val(output_region), path(reference), path(reference_index), path(map)
+        tuple val(meta) , path(input, arity: '1..*'), path(input_index), path(samples_file), val(input_region), val(output_region), path(reference), path(reference_index), path(map)
```
Comment (maxulysse) [inline modules/nf-core/glimpse2/phase/main.nf:21]: <3

File: workflows/phaseimpute/main.nf (hunk 25 above)
Comment (maxulysse) [reply workflows/phaseimpute/main.nf:106]: No idea, that's more a question for you I guess

Comment (maxulysse) [reply workflows/phaseimpute/main.nf:106]: Maybe just create an issue so you remember to think about it later on

**Review by maxulysse — APPROVED** [review]
I love it

**Review by LouisLeNezet — COMMENTED** [review]
Ready to go
